### ISSUE TITLE
LMDB Crashes on Deletion if Keys aren't Sorted

### ISSUE THREAD

**Issue by hraberg** [xtdb/xtdb#2014]
First seemed related to xtdb/xtdb#2009 but was already there. I recently removed the sort for insert, as this was a misguided premature optimisation from way back. Trying to remove it from deletion as well causes LMDB to crash the JVM when closing it.

This obviously sounds like the root cause lies somewhere else. Spent some time debugging this but couldn't figure it out so parked it, put back the sort for now. Crux only deletes things from the index on eviction in practice.

**Comment by hraberg** [comment]
This still happens after switching to buffers, sort still fixes it.

**Comment by hraberg** [comment]
Spent another hour trying to understand what the root cause is here. One new finding is that storing all the keys one by one also crashes. But haven't gotten to the bottom of this. It happens in both LWJGL and lmdbjava, xtdb/xtdb#142.

Occasionally it crashes when it actually tries to store, sometimes when it commits the transaction, and sometimes when it closes the env. I've sanity checked that all keys we use are within the LMDB max key size (but think you will get a different real exception for this). I've also tried to sanity check and copy the buffers extra times in various ways. But nothing seems to help.

**Comment by hraberg** [comment]
One way to take this card forward is to double check all official LMDB documentation to see if we're doing something wrong somewhere, and also sanity check compared to the tests of LMDB itself. We can then also try to reproduce it purely in C:
https://github.com/LMDB/lmdb

**Comment by hraberg** [comment]
The docs to LMDB itself lives here:
http://www.lmdb.tech/doc/

**Comment by hraberg** [comment]
I think the sort is a red herring, have a test that can consistently crash LMDB on env close at 200 long keys, regardless sorting. Not closing LMDB does work, but is obviously not a real solution. But this hints at something is potentially still "open" which one need to ensure isn't before closing?

**Comment by hraberg** [comment]
I've managed to write a test where sorting doesn't help, but short (single key) transactions does. This version doesn't solve the other, original eviction test which still needs sorting. Have no version of the code that handles both.

**Comment by hraberg** [comment]
Does crash on CircleCI as well, so it's not a local issue:
https://circleci.com/gh/juxt/crux/1608

**Comment by hraberg** [comment]
This seems somehow potentially be related to `MDB_WRITEMAP` which I added due to xtdb/xtdb#92 way back. Later I have started to think that the reason for those crashes have more to do with my laptop, so that might have been premature. As mentioned in that issue, one should really keep track of open transactions before resizing the map when using `MDB_WRITEMAP`, but we don't, which is an issue. Though the way these test work I don't think that's the actual issue.

In the LMDB docs it says about this flag: " This is faster and uses fewer mallocs, but loses protection from application bugs like wild pointer writes and other bad updates into the database. Incompatible with nested transactions." So potentially there's still some bug somewhere that causes this flag not to work.

But will turn off this flag for now, as that at least solves this bug. If xtdb/xtdb#92 shows itself again, on relevant machines, we can revisit.
